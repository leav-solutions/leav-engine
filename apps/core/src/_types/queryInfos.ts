// Copyright LEAV Solutions 2017
// This file is released under LGPL V3
// License text available at https://www.gnu.org/licenses/lgpl-3.0.txt
import {IDbProfiler} from './dbProfiler';
import {IValueVersion} from './value';

export interface IQueryInfos {
    userId?: string;
    lang?: string;
    queryId?: string;
    version?: IValueVersion;
    treeId?: string;
    groupsId?: string[];
    applicationId?: string;
    dbProfiler?: IDbProfiler;
}
