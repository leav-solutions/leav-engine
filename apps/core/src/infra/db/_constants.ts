// Copyright LEAV Solutions 2017
// This file is released under LGPL V3
// License text available at https://www.gnu.org/licenses/lgpl-3.0.txt
export const ARANGODB_WRITEWRITE_CONFLICT_CODE = 1200;
export const ARANGODB_HTTP_CONFLICT_CODE = 409;
