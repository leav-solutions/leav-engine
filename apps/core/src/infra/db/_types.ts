// Copyright LEAV Solutions 2017
// This file is released under LGPL V3
// License text available at https://www.gnu.org/licenses/lgpl-3.0.txt
// Copyright LEAV Solutions 2017
// This file is released under LGPL V3

import {AqlQuery} from 'arangojs/lib/cjs/aql-query';
import {IQueryInfos} from '_types/queryInfos';

// License text available at https://www.gnu.org/licenses/lgpl-3.0.txt
export interface IDbDocument {
    _id: string;
    _key: string;
    _rev: string;
    [key: string]: any;
}

export interface IDbEdge extends IDbDocument {
    _from: string;
    _to: string;
}

export interface IExecute {
    query: string | AqlQuery;
    ctx: IQueryInfos;
    withTotalCount?: boolean;
    attempts?: number;
}

export interface IExecuteWithCount<T = IDbDocument> {
    totalCount: number;
    results: T[];
}

export const isExecuteWithCount = (res: IExecuteWithCount<unknown> | unknown[]): res is IExecuteWithCount<unknown> => {
    return typeof (res as IExecuteWithCount).results !== 'undefined';
};
