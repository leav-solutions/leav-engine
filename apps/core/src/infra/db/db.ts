// Copyright LEAV Solutions 2017
// This file is released under LGPL V3
// License text available at https://www.gnu.org/licenses/lgpl-3.0.txt
import {Database} from 'arangojs';
import {IConfig} from '_types/config';

interface IDeps {
    config?: IConfig;
}

const _getDbConnection = (config: IConfig): Database => {
    return new Database({
        url: config.db.url
    });
};

export default function (deps: IDeps): Database {
    const db = _getDbConnection(deps.config);

    return db.database(deps.config.db.name);
}

export const initDb = async (config: IConfig) => {
    const db = _getDbConnection(config);

    const databases = await db.listDatabases();
    const dbExists = databases.reduce((exists, d) => exists || d === config.db.name, false);

    if (!dbExists) {
        await db.createDatabase(config.db.name);
    }

    db.close();
};
