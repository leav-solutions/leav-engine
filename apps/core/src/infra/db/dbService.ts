// Copyright LEAV Solutions 2017
// This file is released under LGPL V3
// License text available at https://www.gnu.org/licenses/lgpl-3.0.txt
import {getCallStack} from '@leav/utils';
import {Database} from 'arangojs';
import {isAqlQuery} from 'arangojs/lib/cjs/aql-query';
import {createHash} from 'crypto';
import {IUtils} from 'utils/utils';
import {IConfig} from '_types/config';
import {IDbProfiler} from '_types/dbProfiler';
import {IDbDocument, IExecute, IExecuteWithCount} from './_types';

const MAX_ATTEMPTS = 10;

export interface IDbService {
    db?: Database;

    /**
     * Execute an AQL query
     * If withTotalCount is set to true, return an object with totalCount and results. Otherwise, just return
     * results straight from DB.
     *
     * @param query
     * @param withTotalCount
     * @param attempts Used when we have to retry a query after a write-write conflict
     * @throws If query fails or we still have a conflict after all attempts
     */
    execute?<T extends IExecuteWithCount<unknown> | unknown[] = IDbDocument[]>(params: IExecute): Promise<T>;

    /**
     * Create a new collection in database
     *
     * @param name Collection name
     * @param type Document or edge collection?
     * @throws If collection already exists
     */
    createCollection?(name: string, type?: collectionTypes): Promise<void> | null;

    /**
     * Delete a collection in database
     *
     * @param name Collection name
     * @param type Document or edge collection?
     * @throws If collection already exists
     */
    dropCollection?(name: string, type?: collectionTypes): Promise<void> | null;

    /**
     * Check if collection already exists
     *
     * @param name
     */
    collectionExists?(name: string): Promise<boolean>;
}

export enum collectionTypes {
    DOCUMENT = 'document',
    EDGE = 'edge'
}

interface IDeps {
    'core.infra.db'?: Database;
    'core.utils'?: IUtils;
    config?: IConfig;
}

export default function ({
    'core.infra.db': db = null,
    'core.utils': utils = null,
    config = null
}: IDeps = {}): IDbService {
    const collectionExists = async function (name: string): Promise<boolean> {
        const collections = await db.listCollections();

        return collections.reduce((exists, c) => exists || c.name === name, false);
    };

    const _sleep = (ms: number): Promise<void> => {
        return new Promise(resolve => setTimeout(resolve, ms));
    };

    return {
        db,
        async execute<T extends IExecuteWithCount<any> | any[] = any[]>({
            query,
            ctx,
            withTotalCount = false,
            attempts = 0
        }: IExecute): Promise<T> {
            try {
                if (config.dbProfiler.enable) {
                    const dbProfiler: IDbProfiler = ctx.dbProfiler ?? {
                        totalCount: 0,
                        uniqueQueriesCount: 0,
                        queries: {}
                    };

                    dbProfiler.totalCount = (dbProfiler.totalCount ?? 0) + 1;

                    // Generate a hash from the query to be able
                    // to group identical queries (exact same query with exact same params)
                    const queryKey = createHash('md5').update(JSON.stringify(query)).digest('base64');

                    if (!dbProfiler.queries) {
                        dbProfiler.queries = {};
                    }

                    const callStack = JSON.stringify(getCallStack(10));

                    dbProfiler.queries[queryKey] = {
                        count: (dbProfiler.queries?.[queryKey]?.count ?? 0) + 1,
                        callers: (dbProfiler.queries?.[queryKey]?.callers ?? new Set()).add(callStack),
                        query
                    };

                    dbProfiler.uniqueQueriesCount = Object.keys(dbProfiler.queries).length;

                    ctx.dbProfiler = dbProfiler;
                }

                // Convert query to AqlQuery if we have a simple query to match query() types
                const queryToRun = isAqlQuery(query)
                    ? {...query}
                    : {
                          query,
                          bindVars: {}
                      };

                const queryOptions = withTotalCount ? {count: true, options: {fullCount: true}} : {};

                const cursor = await db.query(queryToRun, queryOptions);

                const results = await cursor.all();
                return withTotalCount ? {totalCount: cursor.extra.stats.fullCount, results} : results;
            } catch (e) {
                // Handle write-write conflicts: we try the query again with a growing delay between trials.
                // If we reach maximum attempts and still no success, stop it and throw the exception
                // error 1200 === conflict
                if (e.isArangoError && e.errorNum === 1200 && attempts < MAX_ATTEMPTS) {
                    const timeToWait = 2 ** attempts;
                    await _sleep(timeToWait);
                    return this.execute({query, ctx, withTotalCount, attempts: attempts + 1});
                }

                e.message += `\nQuery was: ${JSON.stringify(query).replace(/\\n/g, ' ')}`;
                e.query = query;

                // Response contains circular references which can cause an error when converted to JSON
                // later on. It doesn't contains useful information anyway, so throw it away.
                delete e.response;

                utils.rethrow(e);
            }
        },
        async createCollection(name: string, type = collectionTypes.DOCUMENT): Promise<void> {
            if (await collectionExists(name)) {
                throw new Error(`Collection ${name} already exists`);
            }

            if (type === collectionTypes.EDGE) {
                const collection = db.edgeCollection(name);
                await collection.create();
            } else {
                const collection = db.collection(name);
                await collection.create();
            }
        },
        async dropCollection(name: string, type = collectionTypes.DOCUMENT): Promise<void> {
            if (!(await collectionExists(name))) {
                throw new Error(`Collection ${name} does not exist`);
            }

            const collection = type === collectionTypes.EDGE ? db.edgeCollection(name) : db.collection(name);
            await collection.drop();
        },
        collectionExists
    };
}
