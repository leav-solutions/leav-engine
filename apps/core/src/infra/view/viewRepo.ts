// Copyright LEAV Solutions 2017
// This file is released under LGPL V3
// License text available at https://www.gnu.org/licenses/lgpl-3.0.txt
import {aql} from 'arangojs';
import {IList} from '_types/list';
import {IQueryInfos} from '_types/queryInfos';
import {IGetCoreEntitiesParams} from '_types/shared';
import {IViewRepo, IViewRepoDeps, VIEWS_COLLECTION_NAME} from './_types';

export default function ({
    'core.infra.db.dbService': dbService = null,
    'core.infra.db.dbUtils': dbUtils = null
}: IViewRepoDeps): IViewRepo {
    return {
        async updateView(view: IView, ctx: IQueryInfos): Promise<IView> {
            const collec = dbService.db.collection(VIEWS_COLLECTION_NAME);
            const docToInsert = dbUtils.convertToDoc(view);

            const updatedView = await dbService.execute({
                query: aql`
                    UPDATE ${docToInsert} IN ${collec}
                    OPTIONS {mergeObjects: false, keepNull: false}
                    RETURN NEW`,
                ctx
            });

            return dbUtils.cleanup(updatedView[0]);
        },
        async createView(view: IView, ctx: IQueryInfos): Promise<IView> {
            const collec = dbService.db.collection(VIEWS_COLLECTION_NAME);
            const docToInsert = dbUtils.convertToDoc(view);

            const newView = await dbService.execute({
                query: aql`INSERT ${docToInsert} IN ${collec} RETURN NEW`,
                ctx
            });

            return dbUtils.cleanup(newView[0]);
        },
        async getViews(params: IGetCoreEntitiesParams, ctx: IQueryInfos): Promise<IList<IView>> {
            const defaultParams: IGetCoreEntitiesParams = {
                filters: null,
                strictFilters: false,
                withCount: false,
                pagination: null,
                sort: null
            };
            const initializedParams = {...defaultParams, ...params};

            return dbUtils.findCoreEntity<IView>({
                ...initializedParams,
                collectionName: VIEWS_COLLECTION_NAME,
                customFilterConditions: {
                    created_by: (filterKey, filterVal) => aql`el.${filterKey} == ${filterVal} OR el.shared == true`
                },
                ctx
            });
        },
        async deleteView(viewId: string, ctx: IQueryInfos): Promise<IView> {
            const collec = dbService.db.collection(VIEWS_COLLECTION_NAME);

            const deletedView = await dbService.execute({
                query: aql`REMOVE ${{_key: viewId}} IN ${collec} RETURN OLD`,
                ctx
            });

            return dbUtils.cleanup(deletedView[0]);
        }
    };
}
