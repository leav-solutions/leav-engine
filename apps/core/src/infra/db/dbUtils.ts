// Copyright LEAV Solutions 2017
// This file is released under LGPL V3
// License text available at https://www.gnu.org/licenses/lgpl-3.0.txt
import {aql} from 'arangojs';
import {GeneratedAqlQuery} from 'arangojs/lib/cjs/aql-query';
import {CollectionType} from 'arangojs/lib/cjs/collection';
import {AwilixContainer} from 'awilix';
import {accessSync, constants, readdirSync} from 'fs';
import {IPluginsRepo} from 'infra/plugins/pluginsRepo';
import * as path from 'path';
import * as winston from 'winston';
import {IAttribute} from '_types/attribute';
import {IConfig} from '_types/config';
import {ILibrary} from '_types/library';
import {IList, IPaginationParams, ISortParams} from '_types/list';
import {IQueryInfos} from '_types/queryInfos';
import {IKeyValue} from '_types/shared';
import {ITree} from '_types/tree';
import {IDbValueVersion, IValueVersion} from '_types/value';
import {ECacheType, ICachesService} from '../../infra/cache/cacheService';
import {collectionTypes, IDbService} from './dbService';
import runMigrationFiles from './helpers/runMigrationFiles';
import {IExecuteWithCount} from './_types';

export const MIGRATIONS_COLLECTION_NAME = 'core_db_migrations';

export type CustomFilterConditionsFunc = (
    filterKey: string,
    filterVal: string | boolean | string[],
    strictFilters: boolean
) => GeneratedAqlQuery;

export interface IFindCoreEntityParams {
    collectionName: string;
    filters?: ICoreEntityFilterOptions;
    strictFilters?: boolean;
    withCount?: boolean;
    pagination?: IPaginationParams;
    sort?: ISortParams;
    customFilterConditions?: IKeyValue<CustomFilterConditionsFunc>;
    nonStrictFields?: string[];
    ctx: IQueryInfos;
}

export interface IDbUtils {
    migrate?(depsManager: AwilixContainer): Promise<void>;
    cleanup?<T extends {}>(record: {}): T;
    convertToDoc?(obj: {}): any;
    isCollectionExists?(name: string): Promise<boolean>;
    findCoreEntity?<T extends ICoreEntity>(params: IFindCoreEntityParams): Promise<IList<T>>;
    clearDatabase?(): Promise<void>;
}

interface IDeps {
    'core.infra.db.dbService'?: IDbService;
    'core.infra.cache.cacheService'?: ICachesService;
    'core.utils.logger'?: winston.Winston;
    'core.infra.plugins'?: IPluginsRepo;
    config?: IConfig;
}

export default function ({
    'core.infra.db.dbService': dbService = null,
    'core.infra.cache.cacheService': cacheService = null,
    'core.utils.logger': logger = null,
    'core.infra.plugins': pluginsRepo = null,
    config = null
}: IDeps = {}): IDbUtils {
    /**
     * Create the collections used to managed db migrations
     * This can't be in a migration file because we do have to initialize it somewhere
     *
     */
    async function _initMigrationsCollection(): Promise<void> {
        const collections = await dbService.db.listCollections();

        const colExists = collections.reduce((exists, c) => exists || c.name === MIGRATIONS_COLLECTION_NAME, false);
        if (!colExists) {
            const collection = dbService.db.collection(MIGRATIONS_COLLECTION_NAME);
            await collection.create();
        }
    }

    /**
     * Return the filter's conditions based on key and val supplied.
     *
     * @param filterKey
     * @param filterVal
     * @param bindVars
     * @param index
     * @param strictFilters
     */
    function _getFilterCondition(
        filterKey: string,
        filterVal: string | boolean | string[],
        strictFilters: boolean,
        nonStrictFields?: string[]
    ): GeneratedAqlQuery {
        const queryParts = [];

        // If value is an array (types or formats for example),
        // we call this function recursively on array and join filters with an OR
        if (Array.isArray(filterVal)) {
            if (filterVal.length) {
                const valParts = filterVal.map(val => _getFilterCondition(filterKey, val, strictFilters));
                queryParts.push(aql.join(valParts, ' OR '));
            }
        } else {
            if (filterKey === 'label') {
                // Search for label in any language
                const valParts = config.lang.available.map(l => aql`LIKE(el.label.${l}, ${filterVal}, true)`);
                queryParts.push(aql.join(valParts, ' OR '));
            } else {
                // Filter with a "like" on ID or exact value in other fields
                queryParts.push(
                    (nonStrictFields ?? []).includes(filterKey) && !strictFilters
                        ? aql`LIKE(el.${filterKey}, ${filterVal}, true)`
                        : aql`el.${filterKey} == ${filterVal}`
                );
            }
        }

        return aql.join(queryParts);
    }

    const ret = {
        /**
         * Run database migrations.
         * It takes all files present in migrations folder and run it if it's never been executed before
         *
         * @param depsManager
         */
        async migrate(depsManager: AwilixContainer): Promise<void> {
            await _initMigrationsCollection();
            const ctx: IQueryInfos = {
                userId: config.defaultUserId,
                queryId: 'run-migrations'
            };
            // Load already ran migrations
            const executedMigrations = await dbService.execute<string[]>({
                query: `
                    FOR m IN core_db_migrations
                    RETURN m.file
                `,
                ctx
            });

            const _runMigrationFiles = (files, folder, prefix = null) =>
                runMigrationFiles({
                    files,
                    executedMigrations,
                    migrationsDir: folder,
                    prefix,
                    deps: {depsManager, dbService, logger},
                    ctx
                });

            /*** Core migrations ***/
            // Load migrations files
            const migrationsDir = path.resolve(__dirname, 'migrations');
            const migrationFiles = readdirSync(migrationsDir).filter(file => file.indexOf('.map') === -1);

            await _runMigrationFiles(migrationFiles, migrationsDir);

            /*** Plugins migrations ***/
            const plugins = pluginsRepo.getRegisteredPlugins();
            for (const plugin of plugins) {
                const pluginMigrationFolder = path.resolve(plugin.path + '/infra/db/migrations');

                try {
                    accessSync(pluginMigrationFolder, constants.R_OK);
                } catch (e) {
                    continue;
                }

                const pluginMigrationFiles = readdirSync(pluginMigrationFolder).filter(
                    file => file.indexOf('.map') === -1
                );

                await _runMigrationFiles(pluginMigrationFiles, pluginMigrationFolder, plugin.infos.name);
            }

            /** Clear cache */
            for (const cacheType of Object.values(ECacheType)) {
                cacheService.getCache(cacheType).deleteAll();
            }
        },

        /**
         * Cleanup every system keys from an object coming from database.
         * _key is kept under 'id'
         *
         * @param obj
         * @return any   Cleaned up object
         */
        cleanup(obj: any): any {
            if (obj === null || typeof obj === 'undefined') {
                return null;
            }

            return Object.keys(obj).reduce((newObj: any, key) => {
                if (key === '_key') {
                    newObj.id = obj[key];
                } else if (key[0] !== '_') {
                    newObj[key] = obj[key];
                }

                return newObj;
            }, {});
        },

        /**
         * Convert an object to an object looking like a DB document
         * id is replaced by _key
         *
         * @param obj
         * @return any   DB document compatible object
         */
        convertToDoc(obj: {}): any {
            const newObj: any = {...obj};

            if (typeof newObj.id !== 'undefined') {
                newObj._key = newObj.id;
            }
            delete newObj.id;

            return newObj;
        },
        /**
         * Search core entities (libraries, attributes, trees)
         *
         * @param collectionName
         * @param filters
         * @param strictFilters
         */
        async findCoreEntity<T extends ITree | ILibrary | IAttribute>(
            params: IFindCoreEntityParams
        ): Promise<IList<T>> {
            const {
                collectionName = null,
                filters = null,
                strictFilters = false,
                withCount = false,
                pagination = null,
                sort = null,
                customFilterConditions = {},
                nonStrictFields = ['label', '_key'],
                ctx = {}
            } = params;

            const collec = dbService.db.collection(collectionName);
            const queryParts = [aql`FOR el IN ${collec}`];

            if (filters !== null) {
                const dbFilters = ret.convertToDoc(filters);
                const filtersKeys = Object.keys(dbFilters);

                for (const filterKey of filtersKeys) {
                    const filterVal = dbFilters[filterKey];

                    // Caller can define some custom functions to generate filter condition (like looking on edges to
                    // filter on libraries linked to an attribute). So, if a custom function is define for this filter,
                    // we use it, otherwise we use the standard filters
                    const filterCondsFunc =
                        typeof customFilterConditions[filterKey] !== 'undefined'
                            ? customFilterConditions[filterKey]
                            : _getFilterCondition;
                    const conds = filterCondsFunc(filterKey, filterVal, strictFilters, nonStrictFields);

                    if (conds?.query) {
                        queryParts.push(aql`FILTER`, conds);
                    }
                }
            }

            if (!!sort) {
                const field = sort.field === 'id' ? '_key' : sort.field;
                queryParts.push(aql`SORT el.${field} ${sort.order}`);
            }

            if (!!pagination) {
                queryParts.push(aql`LIMIT ${pagination.offset || 0}, ${pagination.limit}`);
            }

            queryParts.push(aql`RETURN el`);

            const query = aql.join(queryParts);
            const res = await dbService.execute<IExecuteWithCount | any[]>({query, withTotalCount: withCount, ctx});

            const results = !Array.isArray(res) ? res.results : res;

            return {
                totalCount: withCount ? (res as IExecuteWithCount).totalCount : null,
                list: results.map(ret.cleanup)
            };
        },
        convertValueVersionToDb(version: IValueVersion): IDbValueVersion {
            return Object.keys(version).reduce((allVers, treeName) => {
                const id = version[treeName];

                allVers[treeName] = id;

                return allVers;
            }, {});
        },
        async clearDatabase(): Promise<void> {
            // Drop all collections
            const cols = await dbService.db.listCollections();
            for (const col of cols) {
                const colType =
                    col.type === CollectionType.DOCUMENT_COLLECTION ? collectionTypes.DOCUMENT : collectionTypes.EDGE;
                await dbService.dropCollection(col.name, colType);
            }
        }
    };

    return ret;
}
