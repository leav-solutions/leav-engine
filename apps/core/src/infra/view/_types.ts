// Copyright LEAV Solutions 2017 until 2023/11/05, Copyright Aristid from 2023/11/06
// This file is released under LGPL V3
// License text available at https://www.gnu.org/licenses/lgpl-3.0.txt
import {IDbService} from 'infra/db/dbService';
import {IDbUtils} from 'infra/db/dbUtils';
import {IList} from '_types/list';
import {IGetCoreEntitiesParams} from '_types/shared';
import {IQueryInfos} from '../../_types/queryInfos';

export const VIEWS_COLLECTION_NAME = 'core_views';

export interface IViewRepo {
    createView(view: IView, ctx: IQueryInfos): Promise<IView>;
    updateView(view: IView, ctx: IQueryInfos): Promise<IView>;
    getViews(params: IGetCoreEntitiesParams, ctx: IQueryInfos): Promise<IList<IView>>;
    deleteView(viewId: string, ctx: IQueryInfos): Promise<IView>;
}

export interface IViewRepoDeps {
    'core.infra.db.dbService': IDbService;
    'core.infra.db.dbUtils': IDbUtils;
}
