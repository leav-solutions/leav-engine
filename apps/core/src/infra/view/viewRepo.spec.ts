// Copyright LEAV Solutions 2017 until 2023/11/05, Copyright Aristid from 2023/11/06
// This file is released under LGPL V3
// License text available at https://www.gnu.org/licenses/lgpl-3.0.txt
import {Database} from 'arangojs';
import {IDbUtils} from 'infra/db/dbUtils';
import {IQueryInfos} from '_types/queryInfos';
import {mockView} from '../../__tests__/mocks/view';
import viewRepo from './viewRepo';
import {IViewFilterOptions} from '_types/views';

describe('viewRepo', () => {
    const docViewData = {
        ...mockView,
        _key: mockView.id
    };

    const ctx: IQueryInfos = {
        userId: '1'
    };

    const mockDbUtils: Mockify<IDbUtils> = {
        cleanup: jest.fn().mockReturnValue(mockView),
        convertToDoc: jest.fn().mockReturnValue(docViewData),
        findCoreEntity: global.__mockPromise({list: [{...mockView}]})
    };

    const mockDbServ = {
        db: new Database(),
        execute: global.__mockPromise([docViewData])
    };

    beforeEach(() => jest.clearAllMocks());

    describe('createView', () => {
        test('Should create a new view', async () => {
            const repo = viewRepo({
                'core.infra.db.dbService': mockDbServ,
                'core.infra.db.dbUtils': mockDbUtils as IDbUtils
            });

            const newView = await repo.createView({...mockView}, ctx);

            expect(mockDbServ.execute.mock.calls.length).toBe(1);
            expect(typeof mockDbServ.execute.mock.calls[0][0]).toBe('object'); // AqlQuery
            expect(mockDbServ.execute.mock.calls[0][0].query.query).toMatch(/^INSERT/);
            expect(mockDbServ.execute.mock.calls[0][0].query.query).toMatchSnapshot();
            expect(mockDbServ.execute.mock.calls[0][0].query.bindVars).toMatchSnapshot();

            expect(newView).toMatchObject(mockView);
        });
    });

    describe('updateView', () => {
        test('Should update an existing view', async () => {
            const repo = viewRepo({
                'core.infra.db.dbService': mockDbServ,
                'core.infra.db.dbUtils': mockDbUtils as IDbUtils
            });

            const updatedView = await repo.updateView({...mockView}, ctx);

            expect(mockDbServ.execute.mock.calls.length).toBe(1);
            expect(typeof mockDbServ.execute.mock.calls[0][0]).toBe('object'); // AqlQuery
            expect(mockDbServ.execute.mock.calls[0][0].query.query).toMatch(/UPDATE/);
            expect(mockDbServ.execute.mock.calls[0][0].query.query).toMatchSnapshot();
            expect(mockDbServ.execute.mock.calls[0][0].query.bindVars).toMatchSnapshot();

            expect(updatedView).toMatchObject(mockView);
        });
    });

    describe('getViews', () => {
        test("Get shared and user's views", async () => {
            const repo = viewRepo({
                'core.infra.db.dbService': mockDbServ,
                'core.infra.db.dbUtils': mockDbUtils as IDbUtils
            });

            const filters: IViewFilterOptions = {
                created_by: '1',
                library: 'test_lib'
            };

            const views = await repo.getViews({filters}, ctx);

            expect(mockDbUtils.findCoreEntity.mock.calls.length).toBe(1);
            expect(views).toEqual({list: [mockView]});
        });
    });

    describe('deleteView', () => {
        test('Should delete view', async () => {
            const repo = viewRepo({
                'core.infra.db.dbService': mockDbServ,
                'core.infra.db.dbUtils': mockDbUtils as IDbUtils
            });

            const delView = await repo.deleteView(mockView.id, ctx);

            expect(mockDbServ.execute.mock.calls.length).toBe(1);
            expect(typeof mockDbServ.execute.mock.calls[0][0]).toBe('object'); // AqlQuery
            expect(mockDbServ.execute.mock.calls[0][0].query.query).toMatch(/REMOVE/);
            expect(mockDbServ.execute.mock.calls[0][0].query.query).toMatchSnapshot();
            expect(mockDbServ.execute.mock.calls[0][0].query.bindVars).toMatchSnapshot();

            expect(delView).toMatchObject(mockView);
        });
    });
});
