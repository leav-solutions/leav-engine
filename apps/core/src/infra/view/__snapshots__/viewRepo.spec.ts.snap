// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`viewRepo createView Should create a new view 1`] = `"INSERT @value0 IN @@value1 RETURN NEW"`;

exports[`viewRepo createView Should create a new view 2`] = `
Object {
  "@value1": "core_views",
  "value0": Object {
    "_key": "test_view",
    "color": "#123456",
    "created_at": 1234567890,
    "created_by": "1",
    "description": Object {
      "fr": "My test view",
    },
    "display": Object {
      "size": "MEDIUM",
      "type": "list",
    },
    "filters": Array [
      Object {
        "field": "id",
        "value": "fake_id_filter",
      },
    ],
    "id": "test_view",
    "label": Object {
      "fr": "My view",
    },
    "library": "test_lib",
    "modified_at": 1234567890,
    "settings": Object {
      "columns": Array [
        "id",
        "label",
      ],
    },
    "shared": true,
    "sort": Object {
      "field": "id",
      "order": "asc",
    },
  },
}
`;

exports[`viewRepo deleteView Should delete view 1`] = `"REMOVE @value0 IN @@value1 RETURN OLD"`;

exports[`viewRepo deleteView Should delete view 2`] = `
Object {
  "@value1": "core_views",
  "value0": Object {
    "_key": "test_view",
  },
}
`;

exports[`viewRepo updateView Should update an existing view 1`] = `
"
                    UPDATE @value0 IN @@value1
                    OPTIONS {mergeObjects: false}
                    RETURN NEW"
`;

exports[`viewRepo updateView Should update an existing view 2`] = `
Object {
  "@value1": "core_views",
  "value0": Object {
    "_key": "test_view",
    "color": "#123456",
    "created_at": 1234567890,
    "created_by": "1",
    "description": Object {
      "fr": "My test view",
    },
    "display": Object {
      "size": "MEDIUM",
      "type": "list",
    },
    "filters": Array [
      Object {
        "field": "id",
        "value": "fake_id_filter",
      },
    ],
    "id": "test_view",
    "label": Object {
      "fr": "My view",
    },
    "library": "test_lib",
    "modified_at": 1234567890,
    "settings": Object {
      "columns": Array [
        "id",
        "label",
      ],
    },
    "shared": true,
    "sort": Object {
      "field": "id",
      "order": "asc",
    },
  },
}
`;
